### ISSUE TITLE
[Bug]: Speed of the Transcription and its Accuracy

### ISSUE THREAD

**Issue by AdityaYadavalli1** [speechbrain/speechbrain#1616]
### Describe the bug

I am using `EncoderDecoderASR` to run inference on my local model. It takes over 10 minutes to transcribe a single wav file of length 8-9 seconds. Moreover, it gives me transcription that is completely irrelevant to the input wav file. 

Transcription: 'WHAT ARE WE GOING TO DO WHAT ARE WE GOING TO DO WHAT WILL WE DO WHAT WILL WE DO WHAT WILL WE DO WHAT WILL WE DO WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WILL WE DO GOD KNOWS WHAT WOULD WE DO GOD KNOWS WHAT WOULD WE DO GOD KNOWS WHAT WOULD WE DO GOD KNOWS WHAT WOULD WE DO GOD KNOWS WHAT WOULD WE DO GOD WOULD WE DO GOD KNOWS WHAT WOULD WE DO GOD ANY GOD KNOWS GOD KNOWS WHAT GOD KNOWS WHAT GOD IS GOD IS GOD IS GOD KNOWS WHAT GOD IS GOD AND GOD IS GOD IS GOD'

### Expected behaviour

`EncoderDecoderASR` should give me the correct transcription, i.e, 'PLEASE TAKE THE SHAPE OF A LONG ROUND ARCH ...' and should take less than 10 seconds to give me that output. The input is the `example.wav` file that is comes with SpeechBrain by default. 

### To Reproduce

```python

from speechbrain.pretrained import EncoderDecoderASR
asr_model = EncoderDecoderASR.from_hparams(source="save/CKPT+2022-10-21+09-43-33+00", hparams_file="/home/axy327/speechbrain/recipes/LibriSpeech/ASR/transformer/hparams/conformer_inf.yaml", savedir="pretrained_models")
asr_model.transcribe_file("/home/axy327/speechbrain/example.wav")
```

The corresponding yaml code. This is very similar to the LibriSpeech conformer yaml code. I made changes to suit my environment and to suit `EncoderDecoderASR()` after looking at the [huggingface yaml](https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech/blob/main/hyperparams.yaml). 

```yaml
# ############################################################################
# Model: E2E ASR with Transformer
# Encoder: Transformer Encoder
# Decoder: Transformer Decoder + (CTC/ATT joint) beamsearch + TransformerLM
# Tokens: unigram
# losses: CTC + KLdiv (Label Smoothing loss)
# Training: Librispeech 960h
# Authors:  Jianyuan Zhong, Titouan Parcollet 2021
# ############################################################################

# Feature parameters
sample_rate: 16000
n_fft: 400
n_mels: 80

####################### Model parameters ###########################
# Transformer
d_model: 512
nhead: 4
num_encoder_layers: 8
num_decoder_layers: 6
d_ffn: 2048
transformer_dropout: 0.1
activation: !name:torch.nn.GELU
output_neurons: 5000
vocab_size: 5000

# Outputs
blank_index: 0
label_smoothing: 0.1
pad_index: 0
bos_index: 1
eos_index: 2

# Decoding parameters
min_decode_ratio: 0.0
max_decode_ratio: 1.0
valid_search_interval: 10
valid_beam_size: 10
test_beam_size: 66
lm_weight: 0.60
ctc_weight_decode: 0.40

############################## models ################################

CNN: !new:speechbrain.lobes.models.convolution.ConvolutionFrontEnd
    input_shape: (8, 10, 80)
    num_blocks: 3
    num_layers_per_block: 1
    out_channels: (64, 64, 64)
    kernel_sizes: (5, 5, 1)
    strides: (2, 2, 1)
    residuals: (False, False, True)

Transformer: !new:speechbrain.lobes.models.transformer.TransformerASR.TransformerASR
    input_size: 1280
    tgt_vocab: !ref <output_neurons>
    d_model: !ref <d_model>
    nhead: !ref <nhead>
    num_encoder_layers: !ref <num_encoder_layers>
    num_decoder_layers: !ref <num_decoder_layers>
    d_ffn: !ref <d_ffn>
    dropout: !ref <transformer_dropout>
    activation: !ref <activation>
    encoder_module: conformer
    attention_type: RelPosMHAXL
    normalize_before: True
    causal: False

ctc_lin: !new:speechbrain.nnet.linear.Linear
    input_size: !ref <d_model>
    n_neurons: !ref <output_neurons>

seq_lin: !new:speechbrain.nnet.linear.Linear
    input_size: !ref <d_model>
    n_neurons: !ref <output_neurons>

decoder: !new:speechbrain.decoders.S2STransformerBeamSearch
    modules: [!ref <Transformer>, !ref <seq_lin>, !ref <ctc_lin>]
    bos_index: !ref <bos_index>
    eos_index: !ref <eos_index>
    blank_index: !ref <blank_index>
    min_decode_ratio: !ref <min_decode_ratio>
    max_decode_ratio: !ref <max_decode_ratio>
    beam_size: !ref <test_beam_size>
    ctc_weight: !ref <ctc_weight_decode>
    lm_weight: !ref <lm_weight>
    lm_modules: !ref <lm_model>
    temperature: 1.15
    temperature_lm: 1.15
    using_eos_threshold: False
    length_normalization: True

log_softmax: !new:torch.nn.LogSoftmax
    dim: -1

normalizer: !new:speechbrain.processing.features.InputNormalization
    norm_type: global

compute_features: !new:speechbrain.lobes.features.Fbank
    sample_rate: !ref <sample_rate>
    n_fft: !ref <n_fft>
    n_mels: !ref <n_mels>

# This is the Transformer LM that is used according to the Huggingface repository
# Visit the HuggingFace model corresponding to the pretrained_lm_tokenizer_path
# For more details about the model!
# NB: It has to match the pre-trained TransformerLM!!
lm_model: !new:speechbrain.lobes.models.transformer.TransformerLM.TransformerLM
    vocab: 5000
    d_model: 768
    nhead: 12
    num_encoder_layers: 12
    num_decoder_layers: 0
    d_ffn: 3072
    dropout: 0.0
    activation: !name:torch.nn.GELU
    normalize_before: False

tokenizer: !new:sentencepiece.SentencePieceProcessor

Tencoder: !new:speechbrain.lobes.models.transformer.TransformerASR.EncoderWrapper
    transformer: !ref <Transformer>

encoder: !new:speechbrain.nnet.containers.LengthsCapableSequential
    input_shape: [null, null, !ref <n_mels>]
    compute_features: !ref <compute_features>
    normalize: !ref <normalizer>
    cnn: !ref <CNN>
    transformer_encoder: !ref <Tencoder>

# Models
asr_model: !new:torch.nn.ModuleList
    - [!ref <CNN>, !ref <Transformer>, !ref <seq_lin>, !ref <ctc_lin>]

modules:
   compute_features: !ref <compute_features>
   normalizer: !ref <normalizer>
   pre_transformer: !ref <CNN>
   transformer: !ref <Transformer>
   asr_model: !ref <asr_model>
   lm_model: !ref <lm_model>
   encoder: !ref <encoder>
   decoder: !ref <decoder>

# The pretrainer allows a mapping between pretrained files and instances that
# are declared in the yaml.
pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
   loadables:
      normalizer: !ref <normalizer>
      asr: !ref <asr_model>
      lm: !ref <lm_model>
      tokenizer: !ref <tokenizer>
```

### Versions

I am in the default SpeechBrain branch, i.e., develop.
git hash `64c8fa951bdade7e28c81b37452331c61af7b548`

### Relevant log output

```shell
Did not run into any warnings or errors.
```

### Additional context

I have trained the model using only train-clean-100 split with the `conformer.yaml` hparams file in the LibriSpeech recipe and used pretrained LM as in the ASR directory inside `LibriSpeech/transformer`. The model converged giving me WER of **8.23** on test-clean split (understandable given that I didn't train using the complete LibriSpeech data). 

Thank you for providing this toolkit to the community!

**Comment by TParcollet** [comment]
Well, this is very hard to debug from our side. The very first thing to try would be to check the output of the transformer decoder before decoding to see if the predicted sequence makes more sense. This kind of looks like a tokenizer issue to be honest ... is the tokenizer used across all models the same ? (LM + AM)

**Comment by TParcollet** [comment]
One thing to try could be to quickly change the test csv file and add your test file to it. Then evaluate on it. If it works that way, it means that your inference yaml is wrong somewhere ...

**Comment by AdityaYadavalli1** [comment]
Thanks for the response @TParcollet! 

I have used the pretrained tokeniser and LM to train my ASR (just trained the AM like in (`speechbrain/recipes/LibriSpeech/ASR/transformer/`). So I hope tokeniser is not the issue?
 
It's interesting to note that the same code (that uses `EncoderDecoderASR` and my model) works perfectly for librispeech test data. However, it just does not work for this specific `example.wav` file. I checked the sampling rate and encoding of the file. It is 16 KHz and 16 bits per second. I did not get the time yet to check where specifically the code fails. But I am happy to share this `example.wav` file and all the model files with you. Meanwhile, are there any other properties of wav files that I should look to match (from the training phase of librispeech)?

I have also tried inferring the same `example.wav` using the librispeech hugging face model and that is able infer the file perfectly so I think there's nothing wrong with the file.

**Comment by AsuMagic** [comment]
If it is easy for you to check, I'm still curious whether there would be any difference with inference through the test `.csv` as TParcollet suggested. Because IMO, if it works that way but not when using `EncoderDecoderASR`, then it _could_ be the same issue as #1590. It's a bit of a wild guess, but it might be worth ruling out.

**Comment by AsuMagic** [comment]
Ok, now I believe there is a good chance this is the same issue as #1590.

tl;dr of the latest developments: `TransformerASR` with `attention_type: RelPosMHAXL` is apparently broken at inference time. Your hyperparameter file seems to confirm that it is what your model is using.

You could try applying the patch as suggested in #1590 (at least until a PR or new release is done). You should not need to retrain your model; only the `EncoderDecoderASR` code path seems to be broken.

**Comment by AdityaYadavalli1** [comment]
Great! I'll test with the new patch and report my results. Thank you so much @AsuMagic!

**Comment by Adel-Moumen** [comment]
Hello @AdityaYadavalli1, 

Any news with this issue, please? Thanks :-)

**Comment by AdityaYadavalli1** [comment]
Hey @Adel-Moumen! Apologies for not acting on this. Been busy with a few tight deadlines. I'll work on this and report by the end of next week.

**Comment by Adel-Moumen** [comment]
Hello @AdityaYadavalli1, any news regarding this issue please?

**Comment by AdityaYadavalli1** [comment]
Hi @Adel-Moumen, thanks for following up! I just checked with speechbrain version `0.5.14`. It still seems to go into an infinite loop. May I share all the model related files and the corresponding wav file that I am trying to infer with you?

**Comment by Adel-Moumen** [comment]
Yes please. It would help a lot! Thanks :-)

**Comment by AdityaYadavalli1** [comment]
mailed you!

**Comment by Adel-Moumen** [comment]
It fed into my infinite to-do list unfortunately. The model is available here: https://iiitaphyd-my.sharepoint.com/:u:/g/personal/aditya_yadavalli_research_iiit_ac_in/EZ5NSpEbY4RNtvqp1EgT0bMBzKUl4yamsvmTGH-gZTFcRg?e=tnchrj if you want to have a look. As we are closely approaching the release of SpeechBrain 1.0, I may have more time later (December/January) to work on this.

**Comment by AdityaYadavalli1** [comment]
Hi @Adel-Moumen! Thanks for following up. Some settings have changed on my end. Here's the link to all the files: https://iiitaphyd-my.sharepoint.com/:u:/g/personal/aditya_yadavalli_alumni_iiit_ac_in/EZ5NSpEbY4RNtvqp1EgT0bMBSUw0JYcbiqLxskBcNe6tfA?e=tiyWge. Please let me know if you need anything else. Also, more than happy to collaborate with you and send a PR that fixes this!

**Comment by Adel-Moumen** [comment]
> Hi @Adel-Moumen! Thanks for following up. Some settings have changed on my end. Here's the link to all the files: [iiitaphyd-my.sharepoint.com/:u:/g/personal/aditya_yadavalli_alumni_iiit_ac_in/EZ5NSpEbY4RNtvqp1EgT0bMBSUw0JYcbiqLxskBcNe6tfA?e=tiyWge](https://iiitaphyd-my.sharepoint.com/:u:/g/personal/aditya_yadavalli_alumni_iiit_ac_in/EZ5NSpEbY4RNtvqp1EgT0bMBSUw0JYcbiqLxskBcNe6tfA?e=tiyWge). Please let me know if you need anything else. Also, more than happy to collaborate with you and send a PR that fixes this!

Hello @AdityaYadavalli1,

Thank you for sharing the weights again.

I have been running some experiments, and yes, I confirm that we are seeing repeated tokens. I investigated a bit more, and what I found out is that when you are using CTC with this model, the model fails to stop at the right moment. Indeed, when running without CTC decoding, we obtain the following transcription: `THE BIRCH CANOE SLID ON THE SMOOTH PLANKS`, while when using CTC, we obtain `THE BIRCH THE BIRCH THE BIRCH THE BIRCH CANOE SLID ON THE SMOOTH BANKS THE BIRCH CANOE SLIDE ON THE SMOOTH PLANKS`.

Could you please share with me your training YAML file, please? I suspect that you are training without CTC loss (weight set to 0.0 in the training script).

Regarding inference speed, it is linked to the above issue. If you want to make it faster you can remove the CTC module in the Beam Searcher. You will definitely see some speed up. Additionally, you can reduce the beam size, and remove the language model if you really want to maximise the speed up. 

Thanks.

Best,
Adel

**Comment by Adel-Moumen** [comment]
> I believe I've used the one inside the librispeech recipe. This one specifically: [`develop`/recipes/LibriSpeech/ASR/transformer/hparams](https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/ASR/transformer/hparams?rgh-link-date=2024-01-31T13%3A50%3A13Z)

Hmmm... the issue is that one YAML may differ from another. 

I tried with one of our recent released pretrained conformer model (google colab: https://colab.research.google.com/drive/18F1T98mIvP3kQYi8f22GmDcGlzVHnx8F?usp=sharing) and everything works as intended. I would suggest to retrain your conformer model especially since we made some modifications that may impact positively the final performances.

**Comment by AdityaYadavalli1** [comment]
Yeah its been a while and I suspect that many changes have been made since then. I'll retrain and let you know if this issue is till there. Thanks!

**Comment by Adel-Moumen** [comment]
> Yeah its been a while and I suspect that many changes have been made since then. I'll retrain and let you know if this issue is till there. Thanks!

Ok thanks! Let me know what your results are :)

**Comment by Adel-Moumen** [comment]
Hello @AdityaYadavalli1, any news on your side please?

**Comment by AdityaYadavalli1** [comment]
I have lost access to the machine where I first saw this bug and I am now not able to reproduce this. I think this is now fixed by some other PRs meanwhile. Thanks a lot for looking into it! If I face this issue again, I'll re-open this.

**Comment by asumagic** [comment]
Thanks for checking!
